src/Microsoft.AspNet.Mvc.Abstractions/ModelBinding/Validation/ValidationStateDictionary.cs

```diff
@@ -0,0 +1,179 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.AspNet.Mvc.ModelBinding.Validation
+{
+    /// <summary>
+    /// Used for tracking validation state to customize validation behavior for a model object.
+    /// </summary>
+    public class ValidationStateDictionary : 
+        IDictionary<object, ValidationStateEntry>, 
+        IReadOnlyDictionary<object, ValidationStateEntry>
+    {
+        private readonly Dictionary<object, ValidationStateEntry> _inner;
+        
+        /// <summary>
+        /// Creates a new <see cref="ValidationStateDictionary"/>.
+        /// </summary>
+        public ValidationStateDictionary()
+        {
+            _inner = new Dictionary<object, ValidationStateEntry>(ReferenceEqualityComparer.Instance);
+        }
+
+        /// <inheritdoc />
+        public ValidationStateEntry this[object key]
+        {
+            get
+            {
+                ValidationStateEntry entry;
+                TryGetValue(key, out entry);
+                return entry;
+            }
+
+            set
+            {
+                _inner[key] = value;
```

Comment:
In what scenario is a `null` `ValidationStateEntry` interesting? Put another way, why no `null` check here?